### ISSUE TITLE
ImageDataGenerator flow function continue to increase the amount of memory usage

### ISSUE THREAD

**Issue by zhunzhong07** [fchollet/keras#5835]
My keras version is 2.0.1.  When I using the ImageDataGenerator flow function to train the model (train_on_batch), it will continue to increase the amount of memory usage. But the 1.2.0 version does  not appear this problem.

**Comment by joelthchao** [comment]
Do you have a reproducible script to help to identify the problem? For example, same script results memory leak in 2.0.1 but not in 1.2.0.

**Comment by zhunzhong07** [comment]
@joelthchao 
def extract_imagelist(target_path):

     with open(target_path, 'r') as f:
            images, labels = [], []
            for l in f.readlines():
                l = l.strip('\n').split()
                images.append(l[0])
                labels.append(int(l[1]))

     image_len = len(images)
     all_images = np.empty([image_len, 224, 224, 3])
     all_labels = []
     for i in range(image_len):
         img = image.load_img(images[i], target_size=(224, 224))
         img = image.img_to_array(img)
         img = np.expand_dims(img, axis=0)
         img = preprocess_input(img)        
         lab = labels[i]        
         all_images[i, :, :, :] = img
         all_labels.append (lab)
     all_labels = to_categorical(all_labels)
     #yield (all_images, all_labels)
     return all_images, all_labels

files_list = "train.txt"
X, Y = extract_imagelist(files_list)
datagen = ImageDataGenerator(horizontal_flip=True)

net.compile(optimizer=SGD(lr=0.001, momentum=0.9), loss='categorical_crossentropy')

for e in range(10):   
    batches = 0
    for X_batch, Y_batch in datagen.flow(X, Y, batch_size=16):
        loss = net.train_on_batch(X_batch, Y_batch)
        if batches % 20 == 0:
	    print ('Epoch', e, 'iteration:', batches, 'loss:', loss)
        batches += 1
        if batches >= len(X) / 16:
            break

**Comment by joelthchao** [comment]
Are you using theano as backend? May relate to #5935.

**Comment by zhunzhong07** [comment]
@joelthchao No, I use tensorflow as backend.

**Comment by SuJingZhi** [comment]
I have the same issue when I used theano as backend，but once I changed to tensorflow，the issue is gone.

**Comment by abnera** [comment]
@joelthchao thanks for your time on all these issues. You always provide great feedback. As many people have mentioned there seems to be many of us having memory issues when we switch from Keras 1.2 to Keras 2.0. I had to make the switch because of the bug of Batch Normalization + TimeDistributed which was fixed on Keras 2.0. 

Below you have my system specs, and my main fit_generator function. I'm using all using Xception model and keras default `flow_flow_directory` generators. The code works on Keras 1.2 and on a smaller subset of the data without changing a single line of code. I narrow the error to the last batch of training, which means it's related with validation_data `evaluate_generator`. I load a queue of 10k images, this takes about 20/256GB in the computer, but when it's time to validate it fills my RAM 256/256GB + my swap, and system kills the process cuz memory full. On a smaller dataset it works because the validation dataset is not big enough to fill the entire memory. Example training takes like 10GB, but validation takes 40GB on small dataset (20 times smaller than my real dataset). 

Conclusion validating the data it's taking exponentially more resources than my training. Somehow is trying to load all the data into memory even if there is a fix queue size. Or is leaking memory somewhere. If the parameters are exactly the same and it takes 40GB to load the queue in validation why would it take more if the dataset is bigger, unless it's growing the queue to a larger size than specified?

System Specs:
Ubuntu 16.04
CPUs: 44
GPUs: 4
Keras: 2.0.3
Tensorflow: 1.0.1

```
 gpu_model.fit_generator(train_generator,
                            steps_per_epoch=10761,
                            epochs=5,
                            validation_data=validation_generator,
                            validation_steps=3179,
                            max_q_size=10000,
                            workers=40,
                            callbacks=callbacks_list)
```

**Comment by joelthchao** [comment]
@abnera What is the difference between your `train_generator` and `validation_generator`? I need to clarify if the problem comes from setting of generator or validation process itself.

**Comment by abnera** [comment]
@joelthchao good question. I apologize for not making this clear. The main difference is `shuffle=False/True` and The parameters of ImageGenerator. Below is my exact code for them.

```
train_datagen = ImageDataGenerator(rescale=1. / 255,
                                       rotation_range=FLAGS.transformation_ratio,
                                       shear_range=FLAGS.transformation_ratio,
                                       zoom_range=FLAGS.transformation_ratio,
                                       cval=FLAGS.transformation_ratio,
                                       horizontal_flip=True)
    train_generator = train_datagen.flow_from_directory(train_dir,
                                                        target_size=(FLAGS.img_height, FLAGS.img_width),
                                                        batch_size=FLAGS.batch_size,
                                                        class_mode='categorical',
                                                        shuffle=True)

    validation_datagen = ImageDataGenerator(rescale=1. / 255)
    validation_generator = validation_datagen.flow_from_directory(validation_dir,
                                                                  target_size=(FLAGS.img_height, FLAGS.img_width),
                                                                  batch_size=FLAGS.batch_size,
                                                                  class_mode='categorical',
                                                                  shuffle=False)
```

**Comment by joelthchao** [comment]
As you mention, Keras 1.2 has no memory issue. I inspect the [code change](https://github.com/fchollet/keras/compare/1.2.0...2.0.3#diff-93850fa46a789f2e5905894ad0e7bee4) of `ImageDataGenerator` between `1.2.0` and `2.0.3` and I think the memory leak issue is not related to it since there are only style, naming changes. I will continue to check other part of code change.

**Comment by joelthchao** [comment]
I would suggest to do following things to clarify the root cause:
1. Change different models to see if it is model independent.
2. Make train/val generator flow from same data directory to make it data independent.
3. Use same generator for train/val to eliminate configuration issue.
4. Change worker number to clarify if is an issue involve multithreading
5. If possible, switch backend, data_format to see if it is still leaking.
4. Use [callback](https://github.com/fchollet/keras/issues/5935#issuecomment-289041967) to monitor memory usage 

similar issue: #5902

**Comment by abnera** [comment]
@joelthchao thanks for your support. This makes a lot of sense. I literally, just managed to get it to work by using pickle_safe=True. I'll try my best to find the root of the problem to help others, but it takes about 2 hours to run a epoch. I'll update in a couple of days.

**Comment by DBlackKat** [comment]
In Keras 2.0.8 the memory leak issues still occur, I don't have any memory leaks issues before adding image augmentation. My code is similar to @abnera as well, using fit generator and ImageDataGenerator

**Comment by linan7788626** [comment]
Hi there, any updates about this issue? I got a similar memory leak problem in Keras 2.0.8 with either Tensorflow 1.3.0 or Theano 0.9 (Centos 7, CUDA 8.0, cudnn 5.1). If needed, I would provide more details. Any comments or suggestions are highly appreciated.

**Comment by apedevmicrosoft** [comment]
I have the exact same issue, anyone found a solution?

**Comment by alexhernandezgarcia** [comment]
Same issue, which makes it impossible to train with ImageNet

**Comment by Dref360** [comment]
Super late to the party, but for the main issue, the issue is here : 
`for X_batch, Y_batch in datagen.flow(X, Y, batch_size=16):`

You recreate a generator for every epoch, only declare it once and then you'll be allright. 
```
gen =  datagen.flow(X, Y, batch_size=16)
for e in range(10):
    print("MEMORY")
    print(process.memory_info().rss)
    batches = 0
    for X_batch, Y_batch in gen:
        ...
```

**Comment by GlastonburyC** [comment]
You might be super late @Dref360 - but this helped me! :)

**Comment by asutic** [comment]
A similar issue is apparently present in Keras 2.1.6. @Dref360, your suggestion looks reasonable, but in my case I cannot load the whole dataset at a time. It is way too large.  In my code, I break dataset into chunks, load the chunk into the memory, pre-process it, and then try to use ImageDataGenerator's flow() to further augment the data.

So, the problem is most likely coming from the flow() function. What is its behavior when called multiple times?

**Comment by alitk** [comment]
I have the same problem with Keras 2.2.4. It will increase the memory usage each time I call the generator.
`test_datagen  = ImageDataGenerator(preprocessing_function=preprocess_input)
`
I do not know what the problem is and how to free up some memory.

**Comment by universvm** [comment]
I have the same problem with Keras==2.2.2 and tensorflow==1.11.0. I am using this generator https://stanford.edu/~shervine/blog/keras-how-to-generate-data-on-the-fly and have `use_multiprocessing=False` and `workers=1`.

Also I'm on Tensorflow backend. The issue happens when I use training size of .20 which only manages to go for about 200 epochs and then crashes, while .99 goes on for 500 epochs. 

Batch size is 20 images.

Changing type to float as @linan7788626 did not help.

**Comment by alitk** [comment]
I fixed the memory problem with using Python functions. I define a function, and generated the data and run my algorithm on the batch data each time the function is called. Hence after returning from the function, Python removes local variables and hence the memory will be freed up.

**Comment by universvm** [comment]
@alitk I'm using functions within a class (see link above for full code):

```
import numpy as np
import keras

class DataGenerator(keras.utils.Sequence):
    'Generates data for Keras'
    def __init__(self, list_IDs, labels, batch_size=32, dim=(32,32,32), n_channels=1,
                 n_classes=10, shuffle=True):
        'Initialization'
        self.dim = dim
        self.batch_size = batch_size
        self.labels = labels
        self.list_IDs = list_IDs
        self.n_channels = n_channels
        self.n_classes = n_classes
        self.shuffle = shuffle
        self.on_epoch_end()

    def __len__(self):
        'Denotes the number of batches per epoch'
        return int(np.floor(len(self.list_IDs) / self.batch_size))

    def __getitem__(self, index):
        'Generate one batch of data'
        # Generate indexes of the batch
        indexes = self.indexes[index*self.batch_size:(index+1)*self.batch_size]

        # Find list of IDs
        list_IDs_temp = [self.list_IDs[k] for k in indexes]

        # Generate data
        X, y = self.__data_generation(list_IDs_temp)

        return X, y

    def on_epoch_end(self):
        'Updates indexes after each epoch'
        self.indexes = np.arange(len(self.list_IDs))
        if self.shuffle == True:
            np.random.shuffle(self.indexes)

    def __data_generation(self, list_IDs_temp):
        'Generates data containing batch_size samples' # X : (n_samples, *dim, n_channels)
        # Initialization
        X = np.empty((self.batch_size, *self.dim, self.n_channels))
        y = np.empty((self.batch_size), dtype=int)

        # Generate data
        for i, ID in enumerate(list_IDs_temp):
            # Store sample
            X[i,] = np.load('data/' + ID + '.npy')

            # Store class
            y[i] = self.labels[ID]

        return X, keras.utils.to_categorical(y, num_classes=self.n_classes)
```

Not sure where the problem could be

**Comment by aashay96** [comment]
> I fixed the memory problem with using Python functions. I define a function, and generated the data and run my algorithm on the batch data each time the function is called. Hence after returning from the function, Python removes local variables and hence the memory will be freed up.



> I fixed the memory problem with using Python functions. I define a function, and generated the data and run my algorithm on the batch data each time the function is called. Hence after returning from the function, Python removes local variables and hence the memory will be freed up.

Can you show the implementation? I tried the same thing but still had memory issues.

**Comment by skulhare** [comment]
+1
